File: kakao.app/js/ked/edit/editor.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.hasOwn(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, min: function () { var m = Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.min.apply(_k_.min,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n < m ? n : m}}}; return m }, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}}

var editor

import kxk from "../../kxk.js"
let post = kxk.post

import scroll from "../view/base/scroll.js"
import view from "../view/base/view.js"

import gutter from "../view/editor/gutter.js"
import mapview from "../view/editor/mapview.js"

import belt from "./tool/belt.js"

import state from "./state.js"
import draw from "./draw.js"
import complete from "./complete.js"
import mode from "./mode.js"


editor = (function ()
{
    _k_.extend(editor, draw)
    function editor (screen, name, features)
    {
        this["onKey"] = this["onKey"].bind(this)
        this["redraw"] = this["redraw"].bind(this)
        this["onFocus"] = this["onFocus"].bind(this)
        this["hasFocus"] = this["hasFocus"].bind(this)
        this["grabFocus"] = this["grabFocus"].bind(this)
        this["isCursorVisible"] = this["isCursorVisible"].bind(this)
        this["isCursorInEmpty"] = this["isCursorInEmpty"].bind(this)
        this["onWheel"] = this["onWheel"].bind(this)
        this["onMouse"] = this["onMouse"].bind(this)
        this["layout"] = this["layout"].bind(this)
        this["onModesLoaded"] = this["onModesLoaded"].bind(this)
        editor.__super__.constructor.call(this,screen,name,features)
        this.focusable = true
        this.state = new state(this.cells,this.name)
        post.on('focus',this.onFocus)
        if (this.feats.scrllr)
        {
            this.scroll = new scroll(this.screen,this.state,'right')
        }
        if (this.feats.scroll)
        {
            this.scroll = new scroll(this.screen,this.state,'left')
        }
        if (this.feats.mapview)
        {
            this.mapscr = new mapview(this.screen,this.state)
        }
        if (this.feats.gutter)
        {
            this.gutter = new gutter(this)
        }
        if (this.feats.complete)
        {
            this.complete = new complete(this)
        }
        mode.autoStartForEditor(this)
        post.on('modes.loaded',this.onModesLoaded)
        this.onModesLoaded()
    }

    editor.prototype["onModesLoaded"] = function ()
    {
        var m

        var list = _k_.list(mode.names())
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            m = list[_a_]
            if (this.feats[m])
            {
                this.state.allowedModes[m] = true
            }
        }
    }

    editor.prototype["layout"] = function (x, y, w, h)
    {
        var g, m, mw, s, sl, sr, _85_17_

        g = m = s = 0
        sl = sr = 0
        if (this.scroll)
        {
            s = 1
            if (this.feats.scrllr)
            {
                sr = s
                this.scroll.layout(x + w - sr,y,s,h)
            }
            else
            {
                sl = s
                this.scroll.layout(x,y,s,h)
            }
        }
        if (this.gutter)
        {
            g = this.state.gutterWidth()
            this.gutter.layout(x + sl,y,g,h)
        }
        if (this.mapscr)
        {
            if (this.name === 'editor')
            {
                mw = _k_.min(12,parseInt(Math.floor(w / 10)))
            }
            else
            {
                mw = 12
            }
            m = (this.mapscr.visible() ? mw : 0)
            this.mapscr.layout(x + w - sr - m,y,mw,h)
        }
        this.cells.layout(x + sl + g,y,w - s - g - m,h)
        ;(this.complete != null ? this.complete.onEditorLayout() : undefined)
        return this.state.initView()
    }

    editor.prototype["onMouse"] = function (event)
    {
        editor.__super__.onMouse.call(this,event)
    
        var ret, _100_21_, _101_21_, _102_23_, _99_21_

        ret = (this.gutter != null ? this.gutter.onMouse(event) : undefined)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        ret = (this.mapscr != null ? this.mapscr.onMouse(event) : undefined)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        ret = (this.scroll != null ? this.scroll.onMouse(event) : undefined)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        ret = (this.complete != null ? this.complete.onMouse(event) : undefined)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        return false
    }

    editor.prototype["onWheel"] = function (event)
    {
        var inside, res, _117_25_, _118_25_, _119_25_, _123_20_

        if (event.cell[1] >= this.cells.y + this.cells.rows)
        {
            return
        }
        inside = this.cells.isInsideEvent(event)
        inside |= (this.scroll != null ? this.scroll.cells.isInsideEvent(event) : undefined)
        inside |= (this.gutter != null ? this.gutter.cells.isInsideEvent(event) : undefined)
        inside |= (this.mapscr != null ? this.mapscr.cells.isInsideEvent(event) : undefined)
        if (!inside)
        {
            return
        }
        if ((this.complete != null))
        {
            if (res = this.complete.onWheel(event))
            {
                return res
            }
        }
        switch (event.dir)
        {
            case 'up':
            case 'down':
            case 'left':
            case 'right':
                this.state.scrollView(event.dir)
                break
        }

        return {redraw:true}
    }

    editor.prototype["isCursorInEmpty"] = function (cursor)
    {
        cursor = (cursor != null ? cursor : this.state.mainCursor())
        return belt.isLinesPosOutside(this.state.s.lines,cursor)
    }

    editor.prototype["isCursorVisible"] = function (cursor)
    {
        var v

        cursor = (cursor != null ? cursor : this.state.mainCursor())
        v = this.state.s.view
        return (v[0] <= cursor[0] && cursor[0] < v[0] + this.cells.cols) && (v[1] <= cursor[1] && cursor[1] < v[1] + this.cells.rows)
    }

    editor.prototype["grabFocus"] = function ()
    {
        if (this.hidden())
        {
            this.show()
        }
        post.emit('focus',this.name)
        return this.redraw()
    }

    editor.prototype["hasFocus"] = function ()
    {
        return this.state.hasFocus
    }

    editor.prototype["onFocus"] = function (name)
    {
        return this.state.hasFocus = (name === this.name)
    }

    editor.prototype["redraw"] = function ()
    {
        return post.emit('redraw')
    }

    editor.prototype["onKey"] = function (key, event)
    {
        var _190_20_, _194_21_, _199_21_

        if (!this.hasFocus())
        {
            return
        }
        if ((this.complete != null))
        {
            if (this.complete.handleKey(key,event) !== 'unhandled')
            {
                return
            }
        }
        if (this.state.handleKey(key,event) !== 'unhandled')
        {
            ;(this.complete != null ? this.complete.hide() : undefined)
            return
        }
        if (!_k_.empty(event.char))
        {
            this.state.insert(event.char)
            return (this.complete != null ? this.complete.complete() : undefined)
        }
        else
        {
            if (!(_k_.in(key.split('+').slice(-1)[0],['shift','ctrl','alt','cmd'])))
            {
                console.log('editor.onKey?',key)
            }
        }
    }

    return editor
})()

export default editor;

File: kakao.app/love/lua/edit/editor.lua
--[[
    ████████  ███████    ███  █████████   ███████   ████████   
    ███       ███   ███  ███     ███     ███   ███  ███   ███  
    ███████   ███   ███  ███     ███     ███   ███  ███████    
    ███       ███   ███  ███     ███     ███   ███  ███   ███  
    ████████  ███████    ███     ███      ███████   ███   ███  

    configurable text editor
    used by fileeditor, input and choices
    basic editing is handled in state
    complex editing is delegated to mode(s)
--]]

draw = require "edit.draw"
state = require "edit.state"
mode = require "edit.mode"
scroll = require "view.base.scroll"
gutter = require "view.editor.gutter"


local editor = class("editor", draw)
    


function editor:init(name, features) 
        draw.init(self, name, features)
        
        self.name = name
        self.focusable = true
        
        self.state = state(self.cells, self.name)
        
        complete = require "edit.complete"
        
        if self.feats.scrllr then self.scroll = scroll(self.state, 'right') end
        if self.feats.scroll then self.scroll = scroll(self.state, 'left') end
        if self.feats.mapview then self.mapscr = mapview(self.state) end
        if self.feats.gutter then self.gutter = gutter(self) end
        if self.feats.complete then self.complete = complete(self) end
        
        mode.autoStartForEditor(self)
        
        post:on('focus', self.onFocus, self)
        post:on('modes.loaded', self.onModesLoaded, self)
        
        self:onModesLoaded()
        return self
    end


function editor:onModesLoaded() 
        for m in mode.names():each() do 
            if self.feats[m] then 
                self.state.allowedModes[m] = true
            end
        end
    end

-- 000       0000000   000   000   0000000   000   000  000000000  
-- 000      000   000   000 000   000   000  000   000     000     
-- 000      000000000    00000    000   000  000   000     000     
-- 000      000   000     000     000   000  000   000     000     
-- 0000000  000   000     000      0000000    0000000      000     


function editor:layout(x, y, w, h) 
        local g = 0
        local m = 0
        local s = 0
        local sl = 0
        local sr = 0
        
        if self.scroll then 
            s = 1
            if self.feats.scrllr then 
                sr = s
                self.scroll:layout(((x + w) - sr), y, s, h)
            else 
                sl = s
                self.scroll:layout(x, y, s, h)
            end
        end
        
        if self.gutter then 
            g = self.state:gutterWidth()
            self.gutter:layout((x + sl), y, g, h)
        end
        
        if self.mapscr then 
            if (self.name == 'editor') then 
                local mw = min(12, floor((w / 10)))
            else 
                local mw = 12
            end
            
            m = (function () 
    if self.mapscr.visible() then 
    return mw else 
    return 0
                end
end)()
            self.mapscr:layout((((x + w) - sr) - m), y, mw, h)
        end
        
        self.cells:layout(((x + sl) + g), y, (((w - s) - g) - m), h)
        
        if self.complete then 
            self.complete:onEditorLayout()
        end
        
        return self.state:initView()
    end

-- 00     00   0000000   000   000   0000000  00000000  
-- 000   000  000   000  000   000  000       000       
-- 000000000  000   000  000   000  0000000   0000000   
-- 000 0 000  000   000  000   000       000  000       
-- 000   000   0000000    0000000   0000000   00000000  


function editor:onMouse(event) 
        draw.onMouse(self, event)
        
        if (self.gutter and self.gutter:onMouse(event)) then return true end
        if (self.mapscr and self.mapscr:onMouse(event)) then return true end
        if (self.scroll and self.scroll:onMouse(event)) then return true end
        if (self.complete and self.complete:onMouse(event)) then return true end
        
        return false
    end

-- 000   000  000   000  00000000  00000000  000      
-- 000 0 000  000   000  000       000       000      
-- 000000000  000000000  0000000   0000000   000      
-- 000   000  000   000  000       000       000      
-- 00     00  000   000  00000000  00000000  0000000  


function editor:onWheel(event) 
        -- log "EDITOR.ONWHEEL" event
        
        if (event.cell[2] >= (self.cells.y + self.cells.rows)) then return end
        
        local inside = self.cells:isInsideEvent(event)
        inside = (inside or self.scroll.cells:isInsideEvent(event))
        inside = (inside or self.gutter.cells:isInsideEvent(event))
        if self.mapscr then 
            inside = (inside or self.mapscr.cells:isInsideEvent(event))
        end
        
        -- log "EDITOR.ONWHEEL INSIDE" inside
        
        if not inside then return end
        
        if self.complete then 
            local res = self.complete:onWheel(event)
            if res then 
                return res
            end
        end
        
        local steps = 0
        if (event.dir == 'up') or (event.dir == 'down') then steps = event.y
        else steps = event.x
        end
        
        if event.shift then steps = steps * 2 end
        if event.ctrl then steps = steps * 2 end
        if event.alt then steps = steps * 2 end
        
        -- log "EDITOR.ONWHEEL" steps, event.dir
        
        if (event.dir == 'up') or (event.dir == 'down') or (event.dir == 'left') or (event.dir == 'right') then self.state:scrollView(event.dir, steps)
        end
        
        return true
    end

--  0000000  000   000  00000000    0000000   0000000   00000000   
-- 000       000   000  000   000  000       000   000  000   000  
-- 000       000   000  0000000    0000000   000   000  0000000    
-- 000       000   000  000   000       000  000   000  000   000  
--  0000000   0000000   000   000  0000000    0000000   000   000  


function editor:isCursorInEmpty(mc) 
        mc = mc or (self.state:mainCursor())
        
        return belt.isLinesPosOutside(self.state.s.lines, mc)
    end


function editor:isCursorVisible(mc) 
        mc = mc or (self.state:mainCursor())
        
        local v = self.state.s.view
        
        return ((((v[1] <= mc[1]) and (mc[1] <= (v[1] + self.cells.cols))) and (v[2] <= mc[2])) and (mc[2] <= (v[2] + self.cells.rows)))
    end

-- 00000000   0000000    0000000  000   000   0000000  
-- 000       000   000  000       000   000  000       
-- 000000    000   000  000       000   000  0000000   
-- 000       000   000  000       000   000       000  
-- 000        0000000    0000000   0000000   0000000   


function editor:grabFocus() 
        if self:hidden() then self:show() end
        post:emit('focus', self.name)
        return -- @redraw()
    end


function editor:hasFocus() 
        return self.state.hasFocus
    end


function editor:onFocus(name) 
        self.state.hasFocus = (name == self.name)
        return self.state.hasFocus
    end

-- 00000000   00000000  0000000    00000000    0000000   000   000  
-- 000   000  000       000   000  000   000  000   000  000 0 000  
-- 0000000    0000000   000   000  0000000    000000000  000000000  
-- 000   000  000       000   000  000   000  000   000  000   000  
-- 000   000  00000000  0000000    000   000  000   000  00     00  

-- 000   000  00000000  000   000  
-- 000  000   000        000 000   
-- 0000000    0000000     00000    
-- 000  000   000          000     
-- 000   000  00000000     000     


function editor:onKey(key, event) 
        if not self:hasFocus() then return end
        
        if self.complete then 
            if (self.complete:handleKey(key, event) ~= 'unhandled') then return true end
        end
        
        if (self.state:handleKey(key, event) ~= 'unhandled') then 
            self.complete:hide()
            return true
        end
        
        if valid(event.char) then 
            -- log "INSERT #{event.char}"
            self.state:insert(event.char)
            -- log "INSERTED #{event.char}"
            if self.complete then 
                self.complete:complete()
            end
            
            return true
        else 
            local splt = kstr.split(key, "+")
            if not array('shift', 'ctrl', 'alt', 'cmd'):has(splt[#splt]) then 
                return print("editor.onKey? |" .. tostring(key) .. "| " .. tostring(event.char) .. " " .. tostring(valid(event.char)) .. "", event)
            end
        end
    end

return editor